server/src/main/webapp/WEB-INF/rails/webpack/single_page_apps/clicky_pipeline_config.tsx

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2020 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {RoutedSinglePageApp} from "helpers/spa_base";
+import m from "mithril";
+import {PipelineConfigPage} from "views/pages/clicky_pipeline_config/pipeline_config";
+import {EnvironmentVariablesTabContent} from "views/pages/clicky_pipeline_config/tabs/common/environment_variables_tab_content";
+import {GeneralOptionsTabContent} from "views/pages/clicky_pipeline_config/tabs/pipeline/general_options_tab";
+import {MaterialsTabContent} from "views/pages/clicky_pipeline_config/tabs/pipeline/materials_tab_content";
+import {ParametersTabContent} from "views/pages/clicky_pipeline_config/tabs/pipeline/parameters_tab_content";
+import {ProjectManagementTabContent} from "views/pages/clicky_pipeline_config/tabs/pipeline/project_management_tab_content";
+import {StagesTabContent} from "views/pages/clicky_pipeline_config/tabs/pipeline/stages_tab_content";
+import {JobsTabContent} from "views/pages/clicky_pipeline_config/tabs/stage/jobs_tab_content";
+import {StageSettingsTabContent} from "views/pages/clicky_pipeline_config/tabs/stage/stage_settings_tab_content";
+
+class RedirectToGeneralTab extends PipelineConfigPage<any> {
+
+  oninit(vnode: m.Vnode<any, any>) {
+    const pipelineName = this.getMeta().pipelineName;
+    m.route.set("/" + pipelineName + "/general");
+  }
+}
+
+export class PipelineConfigSPA extends RoutedSinglePageApp {
+  constructor() {
+    super({
+            "/": new RedirectToGeneralTab(),
+            "/:pipeline_name/:tab_name": new PipelineConfigPage(
+              new GeneralOptionsTabContent(),
+              new ProjectManagementTabContent(),
+              new MaterialsTabContent(),
+              new StagesTabContent(),
+              new EnvironmentVariablesTabContent(),
+              new ParametersTabContent()
+            ),
+            "/:pipeline_name/:stage_name/:tab_name": new PipelineConfigPage(
+              new StageSettingsTabContent(),
+              new JobsTabContent(),
+              new EnvironmentVariablesTabContent()
+            ),
+            "/:pipeline_name/:stage_name/:job_name/:tab_name": new PipelineConfigPage(new EnvironmentVariablesTabContent())
+          });
```

Comment:
why does the PipelineConfigPage accepts different arguments based on which page is rendered?

To avoid the complexity, can we create the entire pipeline model and pipeline-page always, regardless of the tab we're on?
And in the route setup, we just set the current-tab information, which will render appropriate content on the SPA.

We do not build smaller SPAs for each section of the pipeline config. There is only one SPA for the entire pipeline config. And based on the route, we render the SPA differently.

Reply:
> why does the PipelineConfigPage accepts different arguments based on which page is rendered?

This is done like this because of `RoutedSinglePageApp` implementation. According to this it is  entry point for the page and what to render is decided here it self.  Hence we need to decide the routes and component for the route at this level.

> To avoid the complexity, can we create the entire pipeline model and pipeline-page always, regardless of the tab we're on?

Yes, that is something I had in mind but did not get time to refactor this piece of code. I was thinking of moving the tabs inside the page and based on route return the list of tabs to render. Take out the fetchData at the entry point to about API calls on route change.

> We do not build smaller SPAs for each section of the pipeline config. There is only one SPA for the entire pipeline config. And based on the route, we render the SPA differently.

Agreed but route needs to be setup at this level based on design constraints.